README.md
# k-hubeau

[![Latest Release](https://img.shields.io/github/v/tag/kalisio/k-hubeau?sort=semver&label=latest)](https://github.com/kalisio/k-hubeau/releases)
[![Build Status](https://app.travis-ci.com/kalisio/k-hubeau.svg?branch=master)](https://app.travis-ci.com/kalisio/k-hubeau)

A [Krawler](https://kalisio.github.io/krawler/) based service to download data from French open portal [Hub'Eau](https://hubeau.eaufrance.fr/)

## Description

The **k-hubeau** jobs allow to scrape hydrometric data from the following api: [http://hubeau.eaufrance.fr/page/api-hydrometrie](http://hubeau.eaufrance.fr/page/api-hydrometrie).  The downloaded data are stored in a [MongoDB](https://www.mongodb.com/) database and more precisely in 2 collections:
* the `observations` collection stores the observed data:
  * the water level `H` in meter (m)
  * the water flow `Q` in cubic meter per second (m3/s)
* the `stations` collection stores the data of the stations
  
The project consists in 2 jobs:
* the `stations` job scrapes the stations data according a specific cron expression. By default, every day at midnight.
* the `observations` job scrapes the observations according a specific cron expression. By default every 15 minutes.

## Configuration

### Stations

| Variable | Description |
|--- | --- |
| `DB_URL` | The database URL. The default value is `mongodb://127.0.0.1:27017/hubeau` |
| `DEBUG` | Enables debug output. Set it to `krawler*` to enable full output. By default it is undefined. |

### Observations

| Variable | Description |
|--- | --- |
| `DB_URL` | The database URL. The default value is `mongodb://127.0.0.1:27017/hubeau` |
| `TTL` | The observations data time to live. It must be expressed in seconds and the default value is `604 800` (7 days) | 
| `HISTORY` | The duration of the observations data history the job has to download. It must be expressed in milliseconds and the default value is `86 400 000` (1 day) | 
| `TIMEOUT` | The maximum duration of the job. It must be in milliseconds and the default value is `1 800 000` (30 minutes). |
| `DEBUG` | Enables debug output. Set it to `krawler*` to enable full output. By default it is undefined. |

## Deployment

We personally use [Kargo](https://kalisio.github.io/kargo/) to deploy the service.

## Contributing

Please refer to [contribution section](./CONTRIBUTING.md) for more details.

## Authors

This project is sponsored by 

![Kalisio](https://s3.eu-central-1.amazonaws.com/kalisioscope/kalisio/kalisio-logo-black-256x84.png)

## License

This project is licensed under the MIT License - see the [license file](./LICENSE) for details





CHANGELOG.md
Please refer to GitHub issues for changelog starting from v1.4: https://github.com/kalisio/k-hubeau/issues.

### 1.3.0 (2023-03-09)

##### New Features

*  Upgrade NodeJS to v16 and krawler to latest version (closes [#23](https://github.com/kalisio/k-hubeau/pull/23)) ([ede1255d](https://github.com/kalisio/k-hubeau/commit/ede1255d77d4cf04addd7928aa90f54f543f0a17))
*  Integrated predictions based on machine learning models (closes [#22](https://github.com/kalisio/k-hubeau/pull/22)) ([77c5978a](https://github.com/kalisio/k-hubeau/commit/77c5978a4c27d7d59e003aa6d9db984426396933))

##### Bug Fixes

*  dockerfiles ([e31f3d03](https://github.com/kalisio/k-hubeau/commit/e31f3d03a2f06892e26b9c00c4cd3e5257533731))

##### Other Changes

*  Upgrade NodeJS to v16 and krawler to latest version [#23](https://github.com/kalisio/k-hubeau/pull/23) ([d7e2c19d](https://github.com/kalisio/k-hubeau/commit/d7e2c19d2f823414bb3d54bf39e00c1176cc49c3))
*  Integrate predictions based on machine learning models - updated predictions job to use variable ([#22](https://github.com/kalisio/k-hubeau/pull/22)) ([7f2fd988](https://github.com/kalisio/k-hubeau/commit/7f2fd988a01fb25e43b82e18ed832210227d7c11))
*  Integrate predictions based on machine learning models - updated predictions job to use models path  ([#22](https://github.com/kalisio/k-hubeau/pull/22)) ([a724d550](https://github.com/kalisio/k-hubeau/commit/a724d55036d1c1a80af29bbba1229b00c01fdb78))
*  Integrate predictions based on machine learning models - updated predictions job to use models path  ([#22](https://github.com/kalisio/k-hubeau/pull/22)) ([e70fe253](https://github.com/kalisio/k-hubeau/commit/e70fe25372829617d65b703dfa18cebf97475a20))
*  Integrate predictions based on machine learning models - added predictions job image ([#22](https://github.com/kalisio/k-hubeau/pull/22)) ([a048a938](https://github.com/kalisio/k-hubeau/commit/a048a9382b89b8cfaaedabb503cd8c2ee23e2471))
*  Integrate predictions based on machine learning models ([#22](https://github.com/kalisio/k-hubeau/pull/22)) ([b27dbc48](https://github.com/kalisio/k-hubeau/commit/b27dbc48e411116a9e81249f13bdbf009cb227c3))
*  fixed build badge ([ee002345](https://github.com/kalisio/k-hubeau/commit/ee002345121ace13572626383a454c55af142a3d))
*  update npm release script ([5c7c4d73](https://github.com/kalisio/k-hubeau/commit/5c7c4d73708a7f9d532ac614f8343efb44663612))

### 1.2.0 (2020-10-01)

##### New Features

*  Allow to configure the cron expression (close [#20](https://github.com/kalisio/k-hubeau/pull/20)) ([0494c328](https://github.com/kalisio/k-hubeau/commit/0494c32829e3135e8b4de1d0af98c817ab543f85))

### 1.1.0 (2020-04-29)

##### New Features

*  Upgrade to Node 12 (close [#19](https://github.com/kalisio/k-hubeau/pull/19)) ([8965fff8](https://github.com/kalisio/k-hubeau/commit/8965fff8fe5f0b948247e45a4381eec3181e8453))

##### Bug Fixes

*  Observation data are getting too huge (close [#18](https://github.com/kalisio/k-hubeau/pull/18)) ([ee986846](https://github.com/kalisio/k-hubeau/commit/ee98684603cc2538eebb785be80ea62372c82a49))

#### 1.0.1 (2020-01-25)

##### New Features

*  Increase observation scraping (close [#16](https://github.com/kalisio/k-hubeau/pull/16)) ([f4990880](https://github.com/kalisio/k-hubeau/commit/f4990880501a04387158700a446d6ca784058f4b))

##### Bug Fixes

*  MongoAtlas raises some alerts (close [#17](https://github.com/kalisio/k-hubeau/pull/17)) ([ad17494e](https://github.com/kalisio/k-hubeau/commit/ad17494eae75e687fc3584cf151b60552403d692))
*  The index code_station should be unique in the station collection (close [#15](https://github.com/kalisio/k-hubeau/pull/15)) ([1fdbcbf0](https://github.com/kalisio/k-hubeau/commit/1fdbcbf0851aed6b3134b985f46e8d0fe0796ced))

## 1.0.0 (2019-12-12)

##### New Features

*  Enhance jobs configuration (closed [#14](https://github.com/kalisio/k-hubeau/pull/14)) ([8ed1889e](https://github.com/kalisio/k-hubeau/commit/8ed1889e8e6c7997eecd6378f341ffe1a3a7222a))
*  Add a property 'name' to the station documents (close [#10](https://github.com/kalisio/k-hubeau/pull/10)) ([d87f19db](https://github.com/kalisio/k-hubeau/commit/d87f19db2589e65302d65ac16467c78b49b8e8ca))
*  Observation features do not handle the correct name (closed [#5](https://github.com/kalisio/k-hubeau/pull/5)] ([02f4b763](https://github.com/kalisio/k-hubeau/commit/02f4b763b5ed5e670a043d4073747f9d7174b558))
*  Do not store closed station (closed [#6](https://github.com/kalisio/k-hubeau/pull/6)) ([45b8a010](https://github.com/kalisio/k-hubeau/commit/45b8a010d00ed6e9ad6c06655edb1e0d22b8f72b))

##### Bug Fixes

*  The timeout configuration is inconsistent with the cron configuration (close [#11](https://github.com/kalisio/k-hubeau/pull/11)) ([fe9a7e88](https://github.com/kalisio/k-hubeau/commit/fe9a7e88bb0b256705b6628d5de2a2be854d7e0d))







jobfile-predictions.js
import _ from 'lodash'
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'
import { hooks } from '@kalisio/krawler'

const outputDir = './output'

// Configuration
const dbUrl = process.env.DB_URL || 'mongodb://127.0.0.1:27017/kano'
const baseUrl = process.env.PREDIKT_URL || 'http://localhost:5000/predict'
const modelsPath = process.env.PREDIKT_MODELS_PATH || path.join('..', 'predikt', 'models', 'output', 'water_level_rnn', 'multiple', '24H')
const ttl = parseInt(process.env.TTL) || (7 * 24 * 60 * 60)  // duration in seconds
const timeout = parseInt(process.env.TIMEOUT) || (30 * 60 * 1000) // duration in miliseconds
// Now could be HP_RNN or HP_XGB depending on underlying prediction model
const variable = process.env.VARIABLE || 'HP'
const collection = 'hubeau-observations'
// Read available models
const models = fs.readdirSync(modelsPath)
  .filter(model => fs.lstatSync(path.join(modelsPath, model)).isDirectory())

// Create a custom hook to generate tasks
let generateTasks = (options) => {
  return (hook) => {
    let tasks = []
    models.forEach(model => {
      const code_station = `#${model}`
      tasks.push({
        id: code_station,
        options: {
          url: `${options.baseUrl}/${model}`
        }
      })
    })
    hook.data.tasks = tasks
    return hook
  }
}
hooks.registerHook('generateTasks', generateTasks)

export default {
  id: 'hubeau-predictions',
  store: 'fs',
  options: {
    workersLimit: 2,
    faultTolerant: true,
    timeout: timeout
  },
  taskTemplate: {
    id: 'predictions/<%= taskId %>',
    type: 'http'
  },
  hooks: {
    tasks: {
      before: {
      },
      after: {
        readJson: {},
        writeJsonMemory: {
          hook: 'writeJson',
          key: '<%= id %>',
          store: 'memory'
        },
        apply: {
          function: (item) => {
            const predictionFeature = item.data
            const times = _.get(predictionFeature, 'forecastTime.H', [])
            const values = _.get(predictionFeature, 'properties.H', [])
            let features = []
            _.forEach(times, (time, index) => {
              // Use prediction feature as a template
              let feature = _.pick(predictionFeature, ['type', 'geometry', 'runTime', 'properties.code_station'])
              _.set(feature, 'time', time)
              _.set(feature, `properties.${variable}`, values[index])
              features.push(feature)
            })
            if (features.length > 0) console.log('Found ' + features.length + ' new predictions on station ' + _.get(predictionFeature, 'properties.code_station'))
            item.data = features
          }
        },
        writeMongoCollection: {
          chunkSize: 256,
          collection,
          transform: { unitMapping: { time: { asDate: 'utc' }, runTime: { asDate: 'utc' } } }
        },
        clearData: {}
      }
    },
    jobs: {
      before: {
        createStores: [{
          id: 'memory'
        }, {
          id: 'fs',
          options: {
            path: outputDir
          }
        }],
        connectMongo: {
          url: dbUrl,
          // Required so that client is forwarded from job to tasks
          clientPath: 'taskTemplate.client'
        },
        createMongoCollection: {
          clientPath: 'taskTemplate.client',
          collection,
          indices: [ 
            [{ time: 1 }, { expireAfterSeconds: ttl }], // days in s
            { 'properties.code_station': 1 },
            [{ 'properties.code_station': 1, time: -1 }, { background: true }],
            [{ 'properties.code_station': 1, 'properties.H': 1, time: -1 }, { background: true }],
            { 'properties.H': 1 },
            { geometry: '2dsphere' }
          ],
        },
        generateTasks: {
          baseUrl,
        }
      },
      after: {
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: ['memory', 'fs']
      },
      error: {
        disconnectMongo: {
          clientPath: 'taskTemplate.client'
        },
        removeStores: ['memory', 'fs']
    }
    }
  }
}
